src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,431 @@
+-- hasVertex . Right                   == 'hasRightVertex'
+-- @
+hasVertex :: (Ord a, Ord b) => Either a b -> AdjacencyMap a b -> Bool
+hasVertex (Left x)  = hasLeftVertex x
+hasVertex (Right y) = hasRightVertex y
+
+-- | Check if a graph is empty.
+-- Complecity: /O(1)/ time.
+--
+-- @
+-- isEmpty 'empty'                 == True
+-- isEmpty ('overlay' 'empty' 'empty') == True
+-- isEmpty ('vertex' x)            == False
+-- isEmpty                       == (==) 'empty'
+-- @
+isEmpty :: AdjacencyMap a b -> Bool
+isEmpty (BAM lr rl) = Map.null lr && Map.null rl
+
+-- | The number of vertices in the left part in a graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- leftVertexCount 'empty'         == 0
+-- leftVertexCount 'leftVertex' x  == 1
+-- leftVertexCount 'rightVertex' x == 0
+-- leftVertexCount . 'edges'       == 'length' . 'Data.List.nub' . 'map' 'fst'
+-- @
+leftVertexCount :: AdjacencyMap a b -> Int
+leftVertexCount = Map.size . leftAdjacencyMap
+
+-- | The number of vertices in the right part in a graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- rightVertexCount 'empty'         == 0
+-- rightVertexCount 'rightVertex' x == 1
+-- rightVertexCount 'leftVertex' x  == 0
+-- rightVertexCount . 'edges'       == 'length' . 'Data.List.nub' . 'map' 'snd'
+-- @
+rightVertexCount :: AdjacencyMap a b -> Int
+rightVertexCount = Map.size . rightAdjacencyMap
+
+-- | The number of vertices in a graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- vertexCount 'empty'         == 0
+-- vertexCount 'leftVertex' x  == 1
+-- vertexCount 'rightVertex' x == 1
+-- vertexCount g             == 'leftVertexCount' g + 'rightVertexCount' g
+-- @
+vertexCount :: AdjacencyMap a b -> Int
+vertexCount g = leftVertexCount g + rightVertexCount g
+
+-- | The number of edges in a graph.
+-- Complexity: /O(n)/ time.
+--
+-- @
+-- edgeCount 'empty'      == 0
+-- edgeCount ('vertex' x) == 0
+-- edgeCount ('edge' x y) == 1
+-- edgeCount . 'edges'    == 'length' . 'Data.List.nub'
+-- @
+edgeCount :: AdjacencyMap a b -> Int
+edgeCount = Map.foldr (+) 0 . fmap Set.size . leftAdjacencyMap
+
+-- Adds all edges needed to make the graph undirected.
+-- Complexity: /O(m log n)/.
+--
+-- /Note: this function is for internal use only./
+addReverseEdges :: (Ord a, Ord b) => AM.AdjacencyMap (Either a b) -> AM.AdjacencyMap (Either a b)
+addReverseEdges m = AM.overlay m $ AM.edges [ (v, u) | (u, v) <- AM.edgeList m ]
+
+-- | Constructs a bipartite 'AdjacencyMap' from "Algebra.Graph.AdjacencyMap"
+-- with given part identifiers, adding all needed edges to make the graph
+-- undirected and removing all edges inside one part.
+-- Complexity: /O(m log n)/.
+--
+-- @
+-- 'toBipartite ('Algebra.Graph.AdjacencyMap.empty')                                                           == 'empty'
+-- 'leftAdjacencyMap' (toBipartite ('Algebra.Graph.AdjacencyMap.edge' (Left 1) (Right 2)))                       == Map.'Map.singleton' 1 (Set.'Set.singleton' 2)
```

Comment:
See this comment for how to simplify such examples: https://github.com/snowleopard/alga/pull/207/files#r295996443